File: src/main/scala/us/zoom/data/dfence/providers/snowflake/matchers/PlaybookPatternMatchers.scala
package us.zoom.data.dfence.providers.snowflake.matchers

import org.slf4j.LoggerFactory
import us.zoom.data.dfence.providers.snowflake.models.PlaybookPattern
import us.zoom.data.dfence.sql.ObjectName

class PlaybookPatternMatchers(pattern: PlaybookPattern):

  private val log = LoggerFactory.getLogger(getClass)

  object AccountLevelGrant:
    def unapply(value: String): Boolean =
      val result = value.trim == ""
      if !result then
        log.info(
          "Playbook pattern {} match failed for account-level-grant value {}",
          pattern,
          value
        )
      result

  object AccountLevelObject:
    def unapply(snowObj: String): Boolean =
      val result = pattern match
        case PlaybookPattern(None, _, Some("*"))                                                          => false
        case PlaybookPattern(None, _, Some(objName)) if ObjectName.equalObjectName(snowObj.trim, objName) => true
        case _                                                                                            => false
      if !result then
        log.info(
          "Playbook pattern {} match failed for account-level-object value {}",
          pattern,
          snowObj
        )
      result

  object Database:
    def unapply(snowDb: String): Boolean =
      val result = pattern.dbName match
        case Some("*")                                                            => false
        case Some(dbName) if ObjectName.equalObjectName(snowDb.trim, dbName.trim) => true
        case _                                                                    => false
      if !result then
        log.info(
          "Playbook pattern {} match failed for database value {}",
          pattern,
          snowDb
        )
      result

  object Schema:
    def unapply(snowSchema: String): Boolean =
      val result = pattern.schName match
        case Some("*")                                                                        => true
        case Some(schemaName) if ObjectName.equalObjectName(snowSchema.trim, schemaName.trim) => true
        case None                                                                             => true
        case _                                                                                => false
      if !result then
        log.info(
          "Playbook pattern {} match failed for schema value {}",
          pattern,
          snowSchema
        )
      result

  object Object:
    def unapply(snowObj: String): Boolean =
      val result = pattern.objName match
        case Some(objName) if objName == "*"                                         => true
        case Some(objName) if ObjectName.equalObjectName(snowObj.trim, objName.trim) => true
        case None                                                                    => true
        case _                                                                       => false
      if !result then
        log.info(
          "Playbook pattern {} match failed for object value {}",
          pattern,
          snowObj
        )
      result

File: src/main/java/us/zoom/data/dfence/providers/snowflake/revoke/matchers/PlaybookPatternMatchers.java
package us.zoom.data.dfence.providers.snowflake.revoke.matchers;

import io.vavr.control.Option;
import java.util.function.Function;
import java.util.function.Predicate;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import us.zoom.data.dfence.providers.snowflake.revoke.models.PlaybookPattern;
import us.zoom.data.dfence.sql.ObjectName;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class PlaybookPatternMatchers {

  public static Predicate<PlaybookPattern> global() {
    return p -> true;
  }

  public static Predicate<PlaybookPattern> accountObject(String value) {
    return object(value);
  }

  public static Predicate<PlaybookPattern> database(String value) {
    return database().matches(value);
  }

  public static Predicate<PlaybookPattern> schema(String value) {
    return sch().matches(value);
  }

  public static Predicate<PlaybookPattern> object(String value) {
    return obj().matches(value);
  }

  public static Predicate<PlaybookPattern> noMatch() {
    return p -> false;
  }

  public static SelectedField database() {
    return new SelectedField(PlaybookPattern::dbName);
  }

  public static SelectedField sch() {
    return new SelectedField(PlaybookPattern::schName);
  }

  public static SelectedField obj() {
    return new SelectedField(PlaybookPattern::objName);
  }
  public record SelectedField(Function<PlaybookPattern, Option<String>> value) {

    public Predicate<PlaybookPattern> isEmptyOrWildcard() {
      return isEmpty().or(isWildcard());
    }

    public Predicate<PlaybookPattern> isWildcard() {
      return p -> value.apply(p).map(s -> "*".equals(s.trim())).getOrElse(false);
    }

    public Predicate<PlaybookPattern> isEmpty() {
      return p -> value.apply(p).isEmpty();
    }

    public Predicate<PlaybookPattern> isNotEmpty() {
      return isEmpty().negate();
    }

    public Predicate<PlaybookPattern> isNotWildcard() {
      return isWildcard().negate();
    }

    public Predicate<PlaybookPattern> matches(String currentValue) {
      Predicate<PlaybookPattern> matchesValue =
          p ->
              value
                  .apply(p)
                  .map(s -> ObjectName.equalObjectName(s.trim(), currentValue.trim()))
                  .getOrElse(false);
      return isNotEmpty().and(isNotWildcard()).and(matchesValue);
    }
  }
}
